### PR TITLE
CDPCP-7322. Usersync split FreeIPA user-find into multiple requests

### PR DESCRIPTION
**Pull request by handavid**
The user-find command is nearing limits as our customers scale up. In
particular it may hit an inverting-proxy timeout if the FreeIPA instance
is loaded. This commit adds a feature to split up the user-find into
multiple requests.

This feature is guarded by the CDP_USERSYNC_SPLIT_FREEIPA_USER_RETRIEVAL
entitlement. Code for checking this entitlement has been added.

Usersync checks the entitlement once, at the beginning of the operation and
adds the value to the UserSyncOptions. This keeps all entitlement checks in
the same place.

The FreeIPAClient is updated to take the search string and additional parameters
as arguments. Other code that uses the FreeIPAClient.userFindAll method were
opportunistically optimized to retrieve only a list of UIDs instead of the complete
user records to avoid similar scalability issues.

This feature relies on the fact that all successfully synced users will have their
crn in the title field. The title field is searchable by user-find by default.
The actual split of the search space is performed using the known format of user CRNs.
Each user CRN will contain the substring ":user:" followed by a UUID. The UUID uses
hex characters, so the retrieval of users is split into 16 requests. Machine users
CRNs are not as predictable since they include ":machineUser:<name>/uuid". A
single request is sent to retrieve all machine users, separate from the users.

See detailed description in the commit message.

NOTE: I've added the DO_NOT_MERGE label because this code is dependent upon the Entitlement being added to UMS and promoted first.

### CODE REVIEW

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/FreeIpaUsersStateProvider.java [hunk 1]
```diff
@@ -56,6 +68,31 @@ public UsersState getUsersState(FreeIpaClient freeIpaClient) throws FreeIpaClien
...
+            user.getMemberOfGroup().stream()
+                    .filter(group -> !IPA_UNMANAGED_GROUPS.contains(group))
+                    .forEach(group -> builder.addMemberToGroup(group, user.getUid()));
+        } else {
+            LOGGER.warn("User {} is not a member of any groups.", user.getUid());
+        }
+    }
+
+    private Set<User> getUsers(FreeIpaClient freeIpaClient, boolean splitUserRetrieval)  throws FreeIpaClientException {
+        if (splitUserRetrieval) {
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/FreeIpaUsersStateProvider.java:85]: could you add log, just to see which branch is used

**Review by lacikaaa — CHANGES_REQUESTED** [review]
some small suggestions only, thanks for this change!

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/FreeIpaUsersStateProvider.java (hunk 1 above)
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/FreeIpaUsersStateProvider.java:84]: done

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/FreeIpaUsersStateProvider.java [hunk 2]
```diff
@@ -27,27 +30,36 @@
 public class FreeIpaUsersStateProvider {
     private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaUsersStateProvider.class);
 
+    private static final ImmutableList<String> USER_SEARCH_STRINGS = ImmutableList.<String>builder()
```
[the remark this answers is not in the document]
Comment (handavid) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/FreeIpaUsersStateProvider.java]: done

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java [hunk 3]
```diff
@@ -136,16 +138,31 @@ public Optional<User> userFind(String user) throws FreeIpaClientException {
...
+     * @param additionalParams  additional parameters, e.g., ("all", true) or ("pkey_only", true)
+     * @return Set of found Users
+     */
+    public Set<User> userFindAll(Optional<String> flag, Map<String, Object> additionalParams)
+            throws FreeIpaClientException {
+        List<Object> flags = flag.isEmpty() ? List.of() : List.of(flag.get());
+        ImmutableMap.Builder<String, Object> params = ImmutableMap.<String, Object>builder()
+                .put("sizelimit", 0)
+                .put("timelimit", 0);
+        additionalParams.entrySet().forEach(entry -> params.put(entry.getKey(), entry.getValue()));
```
[the remark this answers is not in the document]
Comment (handavid) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: done.

File: freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java [hunk 4]
```diff
@@ -136,16 +138,31 @@ public Optional<User> userFind(String user) throws FreeIpaClientException {
...
+
+    /**
+     * Finds all users in FreeIPA. This command sends a user_find request with the sizelimit
+     * and timelimit set to 0.
+     *
+     * @param flag              the search string
+     * @param additionalParams  additional parameters, e.g., ("all", true) or ("pkey_only", true)
+     * @return Set of found Users
+     */
+    public Set<User> userFindAll(Optional<String> flag, Map<String, Object> additionalParams)
```
[the remark this answers is not in the document]
Comment (handavid) [reply freeipa-client/src/main/java/com/sequenceiq/freeipa/client/FreeIpaClient.java]: done

**Review by lacikaaa — APPROVED** [review]

**Pull request merged**